wallet/createtx.go

```diff
@@ -368,25 +368,38 @@ func (w *Wallet) txToOutputs(outputs []*wire.TxOut, account uint32, minconf int3
 		}
 		pool = w.addrPools[account].internal
 	}
+	changeAddrUsed := false
 	txSucceeded := false
 	pool.mutex.Lock()
 	defer pool.mutex.Unlock()
 	defer func() {
-		if txSucceeded {
+		if txSucceeded && changeAddrUsed {
 			pool.BatchFinish()
 		} else {
 			pool.BatchRollback()
 		}
 	}()
 
-	inputSource := w.TxStore.MakeInputSource(account, minconf, bs.Height)
+	// The change address is pulled here rather than after
+	// MakeInputSource is called because MakeInputSource
+	// accesses the database in a way that is unsafe from
```

Comment:
It's not unsafe, it just deadlocks.  Just say that the two functions can use separate transactions, one an update, and so the update needs to happen first.